File: euler-443-gcd-dequence.kt/euler-443-gcd-dequence.kt
class EulerProblem {
    // calculates the greatest common divisor (GCD) of two numbers
    private fun gcd(a: Int, b: Int): Int {
        var x = a
        var y = b
        while (y != 0) {
            val temp = y
            y = x % y
            x = temp
        }
        return x
    }

    // recursively calculates g(n) based on the problem statement
    private fun g(n: Int): Int {
        return if (n == 4) {
            13
        } else {
            val prevG = g(n - 1)
            prevG + gcd(n, prevG)
        }
    }

    // solves the Euler problem for a given value of n
    fun solveEulerProblem(n: Int): Int {
        return g(n)
    }
}

fun main(args: Array<String>) {
    // Check if there is an argument provided
    if (args.size != 1) {
        println("Usage: kotlin MainKt <n>")
        return
    }

    // Convert the argument to an integer
    val n = args[0].toIntOrNull()
    if (n == null || n <= 0) {
        println("Invalid input. Please provide a valid positive integer for n.")
        return
    }

    val eulerProblem = EulerProblem()
    val result = eulerProblem.solveEulerProblem(n)
    println("$result")
}

File: euler-443-gcd-dequence.go/euler-443-gcd-dequence.go
package main // package declaration for the main package

import (
	"fmt"
	"os"
	"strconv"
)

// function gcd calculates the greatest common divisor of two numbers
func gcd(a, b int) int { // function declaration for calculating GCD
	for b != 0 { // start of a loop to find GCD
		a, b = b, a%b // Euclidean algorithm to find GCD
	}
	return a // return the GCD
}

// function g computes the value of g(n) based on the provided definition
func g(n int) int { // function to calculate g(n)
	if n == 4 { // check if n is equal to 4
		return 13 // return 13 as defined in the problem
	}

	prevG := g(n - 1) // recursively call g function
	return prevG + gcd(n, prevG) // calculate g(n) using the defined formula
}

func main() { // main function
	var input int
	var err error
	input, err = strconv.Atoi(os.Args[1]) // input from command line
	if err != nil {
		// if there is conversion to integer error, exit
		panic(err)
	}
	result := g(input) // calculate g(n)
	fmt.Printf("%d\n", result) // print the value of g(n)
}
